message_ix_models/model/material/demand/demand.py

```diff
@@ -1,42 +1,34 @@
 import logging
 from pathlib import Path
-from typing import Literal, Union
+from typing import TYPE_CHECKING, Callable, List, Literal, Union
 
-import message_ix
 import numpy as np
 import pandas as pd
+import scipy.optimize as opt
 import yaml
 from message_ix import make_df
 from scipy.optimize import curve_fit
 
 import message_ix_models.util
 from message_ix_models import Context, ScenarioInfo
 from message_ix_models.model.material.data_util import get_ssp_soc_eco_data
+from message_ix_models.model.material.demand.config import FittingConfig
```

Comment:
It's good to use submodules where appropriate.

In this case, the directory message_ix_models/model/material/demand/ lacks an `__init__.py` file, and imports are repetitious (`from message_ix_models.model.material.demand.demand import …`).
I would suggest:
- Rename `demand.py` to `__init__.py`.
- Adjust the import on this line to a relative import:
  ```py
  from .config import Fitting Config
  ```
  etc.
- Adjust other imports and references in e.g. the docs.